### PR TITLE
Victorhu/fileversion

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Inject the file version attribute in the generated assembly, only if the attribute is not already present.

Goes together with https://github.com/aspnet/Universe/pull/192 and fixes https://github.com/aspnet/dnx/issues/1504

Please review: @davidfowl, @anurse, @pranavkm, @ChengTian 
Please note this is for beta4/release branch

### CODE REVIEW

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 1]
```diff
@@ -214,17 +214,23 @@ private CompilationModules GetCompileModules(ILibraryKey target)
             CSharpParseOptions parseOptions)
         {
             var emptyVersion = new Version(0, 0, 0, 0);
+            var assemblyAttributes = compilation.Assembly.GetAttributes();
 
-            // If the assembly version is empty then set the version
-            if (compilation.Assembly.Identity.Version == emptyVersion)
+            var addAttributeIfNotExists = new Action<string, string>((attributeName, attributeValue) =>
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Why not make this a regular method?

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 2]
```diff
@@ -188,6 +190,13 @@ internal static Project GetProjectFromStream(Stream stream, string projectName,
                 }
             }
 
+            var fileVersion = Environment.GetEnvironmentVariable("DNX_ASSEMBLY_FILE_VERSION");
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime/Project.cs]: No love for `DNX_ASSEMBLY_INFORMATIONAL_VERSION`?

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs (hunk 1 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Less arguments :)

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 3]
```diff
@@ -62,6 +62,8 @@ public string ProjectDirectory
 
         public SemanticVersion Version { get; private set; }
 
+        public string AssemblyFileVersion { get; private set; }
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime/Project.cs]: Should this be of type `Version`?

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 4]
```diff
@@ -214,17 +214,23 @@ private CompilationModules GetCompileModules(ILibraryKey target)
             CSharpParseOptions parseOptions)
         {
             var emptyVersion = new Version(0, 0, 0, 0);
+            var assemblyAttributes = compilation.Assembly.GetAttributes();
 
-            // If the assembly version is empty then set the version
-            if (compilation.Assembly.Identity.Version == emptyVersion)
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Did this check get removed?

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Yes. Because we already check the attributes

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 3 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: Sure, why not

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 2 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: Nope, no love for it :) That value is automatically computed

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: That value is taken from the `DNX_BUILD_VERSION` variable

**Comment by victorhurdugaci** [comment]
@pranavkm updated the PR. please take a look

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 5]
```diff
@@ -188,6 +190,21 @@ internal static Project GetProjectFromStream(Stream stream, string projectName,
...
+                fileVersion = project.Version.ToString();
+            }
+
+            try
+            {
+                project.AssemblyFileVersion = new Version(fileVersion);
+            }
+            catch(FormatException ex)
+            {
+                throw new FormatException("The assembly file version is invalid: " + fileVersion, ex);
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Project.cs]: Nah. Just use Version.Parse and let that throw.

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 6]
```diff
@@ -188,6 +190,21 @@ internal static Project GetProjectFromStream(Stream stream, string projectName,
                 }
             }
 
+            var fileVersion = Environment.GetEnvironmentVariable("DNX_ASSEMBLY_FILE_VERSION");
+            if (string.IsNullOrWhiteSpace(fileVersion))
+            {
+                fileVersion = project.Version.ToString();
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Project.cs]: Why to ToString and new Verison right underneath? If you're defaulting to the project.Version then just assign it directly.

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 7]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
         private static CSharpCompilation ApplyVersionInfo(CSharpCompilation compilation, Project project,
             CSharpParseOptions parseOptions)
         {
+            const string assemblyFileVersionName = "System.Reflection.AssemblyFileVersionAttribute";
+            const string assemblyVersionName = "System.Reflection.AssemblyVersion";
+            const string assemblyInformationalVersion = "System.Reflection.AssemblyInformationalVersion";
+
             var emptyVersion = new Version(0, 0, 0, 0);
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: You should be able to remove this

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 8]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
...
+            var assemblyAttributes = compilation.Assembly.GetAttributes();
+
+            bool foundAssemblyFileVersion = false;
+            bool foundAssemblyVersion = false;
+            bool foundAssemblyInformationalVersion = false;
 
-            // If the assembly version is empty then set the version
-            if (compilation.Assembly.Identity.Version == emptyVersion)
+            foreach(var assembly in assemblyAttributes)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: spacing

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 9]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
...
             {
-                return compilation.AddSyntaxTrees(new[]
+                string attributeName = assembly.AttributeClass.ToString();
+
+                if (attributeName == assemblyFileVersionName)
                 {
-                    CSharpSyntaxTree.ParseText("[assembly: System.Reflection.AssemblyVersion(\"" + project.Version.Version + "\")]", parseOptions),
-                    CSharpSyntaxTree.ParseText("[assembly: System.Reflection.AssemblyInformationalVersion(\"" + project.Version + "\")]", parseOptions)
+                    foundAssemblyFileVersion = true;
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Why not just add the syntax trees inline instead of having a bool? You can't have more than one of those attributes anyways.

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs (hunk 7 above)
Comment (Eilon) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Also, a more typical empty version is to call the parameterless ctor (so that it has no build/revision in it, as opposed to explicitly 0).

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 10]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
...
             CSharpParseOptions parseOptions)
         {
+            const string assemblyFileVersionName = "System.Reflection.AssemblyFileVersionAttribute";
+            const string assemblyVersionName = "System.Reflection.AssemblyVersion";
+            const string assemblyInformationalVersion = "System.Reflection.AssemblyInformationalVersion";
+
             var emptyVersion = new Version(0, 0, 0, 0);
+            var assemblyAttributes = compilation.Assembly.GetAttributes();
+
+            bool foundAssemblyFileVersion = false;
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: `var`

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime/Project.cs]: Seems like a simple enough change - but we could do it as a separate PR.

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 11]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
...
+                {
+                    foundAssemblyInformationalVersion = true;
+                }
+            }
+
+            if (!foundAssemblyFileVersion)
+            {
+                compilation = compilation.AddSyntaxTrees(new[]
+                {
+                    CSharpSyntaxTree.ParseText("[assembly: " +  assemblyFileVersionName + "(\"" + project.AssemblyFileVersion + "\")]", parseOptions)
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Embrace the C# 6.0 Luke.
`CSharpSyntaxTree.ParseText($"[assembly: {assemblyFileVersionName}(\"${project.AssemblyFileVersion}\")]", parseOptions)`

File: src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs [hunk 12]
```diff
@@ -213,18 +213,59 @@ private CompilationModules GetCompileModules(ILibraryKey target)
...
+            bool foundAssemblyInformationalVersion = false;
 
-            // If the assembly version is empty then set the version
-            if (compilation.Assembly.Identity.Version == emptyVersion)
+            foreach(var assembly in assemblyAttributes)
             {
-                return compilation.AddSyntaxTrees(new[]
+                string attributeName = assembly.AttributeClass.ToString();
+
+                if (attributeName == assemblyFileVersionName)
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: Maybe do `string.Equals(attributeName, assemblyFileVersionName, StringComparison.Ordinal)` here. Also is the value guaranteed to be an fully qualified name all the time? Even if the actual code didn't fully qualify it?

```C#
using System.Reflection;
[assembly: AssemblyVersion("1.0.0")]
```

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime.Roslyn/RoslynCompiler.cs]: > Maybe do  ... 

Ok

> Also is the value guaranteed to be an fully qualified name all the time? Even if the actual code didn't fully qualify it?

Good question and the answer is yes, it is always the full qualified name.

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 2 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: We can change that in the `dev` branch

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 5 above)
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: Nah. The error message is really bad. It just says that the string has an invalid format but doesn't tell you what the value is.

Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/Project.cs]: Then at least fix the formatting in the file :smile:

**Comment by victorhurdugaci** [comment]
Updated

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 13]
```diff
@@ -188,6 +190,23 @@ internal static Project GetProjectFromStream(Stream stream, string projectName,
...
+            var fileVersion = Environment.GetEnvironmentVariable("DNX_ASSEMBLY_FILE_VERSION");
+            if (string.IsNullOrWhiteSpace(fileVersion))
+            {
+                project.AssemblyFileVersion = project.Version.Version;
+            }
+            else
+            {
+                try
+                {
+                    project.AssemblyFileVersion = new Version(fileVersion);
```
Comment (pranavkm) [inline src/Microsoft.Framework.Runtime/Project.cs]: Maybe `TryParse` instead of try-catching?

**Comment by pranavkm** [comment]
:shipit: once somebody from DNX signs off.

File: src/Microsoft.Framework.Runtime/Project.cs [hunk 14]
```diff
@@ -188,6 +190,23 @@ internal static Project GetProjectFromStream(Stream stream, string projectName,
...
+                }
+                catch (FormatException ex)
+                {
+                    throw new FormatException("The assembly file version is invalid: " + fileVersion, ex);
```
Comment (muratg) [inline src/Microsoft.Framework.Runtime/Project.cs]: Consider using String.Format instead of concat.

**Comment by muratg** [comment]
/cc: @joeloff 

:shipit:

File: src/Microsoft.Framework.Runtime/Project.cs (hunk 14 above)
Comment (moozzyk) [inline src/Microsoft.Framework.Runtime/Project.cs]: Doesn't the exception message have to be localizable?

Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Project.cs]: No strings in the runtime are localizable. I think there is an issue about that somewhere...

**Comment by victorhurdugaci** [comment]
Updated the PR with one change: the prefix for the version is picked up from the `project.json` file.
